paymentsheet/src/main/java/com/stripe/android/paymentsheet/PaymentOptionsViewModel.kt

```diff
@@ -242,24 +251,32 @@ internal class PaymentOptionsViewModel @Inject constructor(
     }
 
     /**
-     * If the screen was launched with Link as the previously selected payment and that selection is kept,
-     * make sure the underlying default Link payment method is preserved.
+     * - Updates the [PaymentSelection], if Link, to include the current [LinkAccount] if it exists.
+     * - Preserves the previously selected payment method, if any, in case none is selected in this launch.
      */
-    private fun PaymentSelection.preserveLinkPaymentIfNeeded(): PaymentSelection {
-        return if (this is Link) {
-            val previousLinkPayment = (args.state.paymentSelection as? Link)?.selectedPayment
-            copy(selectedPayment = selectedPayment ?: previousLinkPayment)
-        } else {
-            this
+    private fun PaymentSelection.withLinkDetails(): PaymentSelection = when (this) {
+        is Link -> when (val linkAccount = linkAccountHolder.linkAccount.value) {
+            // If link account is null, clear account status and selected payment from payment selection
+            null -> copy(
+                linkAccount = null,
+                selectedPayment = null
+            )
+            // If link account exists, include it in the payment selection and keep the previously selected payment.
+            else -> copy(
+                linkAccount = linkAccount,
+                selectedPayment = (selectedPayment ?: (args.state.paymentSelection as? Link)?.selectedPayment)
+            )
         }
+        else -> this
     }
```

Comment:
enhances the previous logic to clear the payment method / link account if user logs out, but keeps preserving the selected payment method if user still logged + no new payment method selected.